editoast: core_task: add `SimulationEnv`

leovalais opened the pull request:
> [!NOTE]
> Question: is there a reason to represent power restrictions and margins differently?

> [!TIP]
> 1. The last 6 commits are divided like so to help with the review but will be squashed before merging. They do not work individually and some definitions used in a given commit may be introduced in a later one. Looking at the PR as a whole is likely useful as well.
> 2. Looking at the tests helps. 
> 3. I'm available to peer review as many times as necessary. ◀️ 

> [!CAUTION]
> 1. Some functions or objects with similar purposes are named differently in `SimulationEnv` and `PathfindingEnv`. I find naming in `SimulationEnv` better: if everyone agrees, I'll make a separate renaming PR for `PathfindingEnv` to be symmetric.
> 2. A lot of lines, true, but to be fair almost half of them are documentation and tests 😬 I tried to document and divide as much as possible, moving all refactoring work to other PRs, but now the diff is basically just insertions...

---

* Adds the new `SimulationEnv` responsible for streaming simulation
results.
* Relies on `PathfindingEnv` to fetch paths to build the
simulation requests. 
* Follows the same structure as `PathfindingEnv`,
but the `run` function is left for later. 
* A few naming disparities
with `PathfindingEnv`. If they are accepted, `PathfindingEnv` naming
will be updated accordingly.

Two private submodules for organization:

`mod inputs`
-----

* Simulation parameters and consist, just like `PathfindingEnv`
* Electrical profiles are handled globally

Core requests for margins and power restrictions to be represented
differently, even though they're both essentially values between two
waypoints. The easiest way to represent that is by using a RangeMap,
making construction and iteration easy, coealescing adjacent ranges and
flattening overlapping ranges. All that simplifies both our code and
reduces the payload size. We also avoid relying on path item IDs that
only make sense in the RailJSON (present in the DB table as well), further
separating input schemas and high level representations of our objects.

Is also defined a `SimulationTrain` that ensures that pathfinding
and simulation parameters are consistent. It also helps with building
power restrictions and margins RangeMaps.

`mod request`
-----

Isolates the construction of the request requiring both the simulation
input described there above, and a path, provided by `PathfindingEnv`.

Most of the module is actually for testing the construction, even though
it's a fairly simple function thanks to the RangeMap representation.

leovalais committed:
editoast: common: add utils to impl Eq for floats

For physical quantities, NaN isn't a valid value so we can enforce Eq
for convenience. This notably allows using such structures as HashMap
keys.

Signed-off-by: Leo Valais <leo.valais97@gmail.com>

leovalais committed:
editoast: core_task: add `SimulationEnv`

* Adds the new `SimulationEnv` responsible for streaming simulation
results.
* Relies on `PathfindingEnv` to fetch paths to build the
simulation requests.
* Follows the same structure as `PathfindingEnv`,
but the `run` function is left for later.
* A few naming disparities
with `PathfindingEnv`. If they are accepted, `PathfindingEnv` naming
will be updated accordingly.

Two private submodules for organization:

`mod inputs`
-----

* Simulation parameters and consist, just like `PathfindingEnv`
* Electrical profiles are handled globally

Core requests for margins and power restrictions to be represented
differently, even though they're both essentially values between two
waypoints. The easiest way to represent that is by using a RangeMap,
making construction and iteration easy, coealescing adjacent ranges and
flattening overlapping ranges. All that simplifies both our code and
reduces the payload size. We also avoid relying on path item IDs that
only make sense in the RailJSON (present in the DB table as well), further
separating input schemas and high level representations of our objects.

Is also defined a `SimulationTrain` that ensures that pathfinding
and simulation parameters are consistent. It also helps with building
power restrictions and margins RangeMaps.

`mod request`
-----

Isolates the construction of the request requiring both the simulation
input described there above, and a path, provided by `PathfindingEnv`.

Most of the module is actually for testing the construction, even though
it's a fairly simple function thanks to the RangeMap representation.

Signed-off-by: Leo Valais <leo.valais97@gmail.com>

leovalais commented:
@MartinBourbier this PR isn't the easiest nor the quickest to review to say the least, don't hesitate to ask for a peer review (alongside @woshilapin probably?)

woshilapin commented on editoast/core_task/src/envs/simulation/inputs.rs:
```diff
@@ -0,0 +1,266 @@
+
+/// Schedule information for a path waypoint
+#[derive(Debug, Clone, Hash, PartialEq, Eq)]
+pub enum SimulationWaypointSchedule {
```
Isn’t it the same `struct` as `SimulationScheduleItem` or `ScheduleItem`? Why not reusing it? Especially, the `PassBy` seems to be semantically the same as `stop_for: None`, and also, `ScheduleItem` uses `PositiveDuration` instead of `u64` (we might even think of using `uom::Time`?).

leovalais replied:
Both have extra fields that are not necessary here. I've made `stop_for` not optional as we discussed this afternoon. 2371c7486789fc6a522c9b86fc28c2f39ab7c2d1

woshilapin commented on editoast/core_task/src/envs/simulation/inputs.rs:
```diff
@@ -0,0 +1,267 @@
+    ) -> Self {
+        if !parameters.schedule.is_empty()
+            || !parameters.power_restrictions.is_empty()
+            || !parameters.margins.is_empty()
```
Maybe this can all be regroup into a `SimulationTrainParameters::is_empty()` function?

leovalais replied:
22740e37002a66d11bb959e653344c989fcc3e8e

woshilapin commented on editoast/core_task/src/envs/simulation/inputs.rs:
```diff
@@ -0,0 +1,267 @@
+        let n = self.path_constraints.path_items.len();
+        debug_assert_eq!(n, self.parameters.schedule.len());
+        debug_assert!(
+            self.parameters
+                .power_restrictions
+                .iter()
+                .map(|e| e.0.end)
+                .max()
+                <= Some(n)
```
Not trivial to understand that assertion. It feels like we’re comparing indexes (`n`) with offsets (`.end`) 🤔 Maybe a comment to explain (or a second argument to `debug_assert!()`).

leovalais replied:
Here e.0 is a range of indexes. So 'end' refers to an index meaning the operation is homogeneous. I'll rename the variable to make it more obvious.

leovalais replied:
1c4aecc715f7fd9a23a40e67998b16ce4405022c

woshilapin commented on editoast/core_task/src/envs/simulation/inputs.rs:
```diff
@@ -0,0 +1,266 @@
+/// Use [SimulationTrain] to ensure consistency with the pathfinding constraints.
+#[derive(Debug, educe::Educe)]
+#[educe(Hash, PartialEq, Eq)]
+pub struct SimulationTrainParameters {
```
It seems that this `struct` needs to maintain an invariant between the 3 first fields: `schedule`, `power_restrictions` and `margins`. Maybe avoiding `pub` fields and introducing a constructor and some setters would be a good idea?

leovalais replied:
I must admit I'm not a huge fan since it's verbose and visibility was already narrowed, but why not if it helps the reader have better confidence about the invariant 83d36fa5c142decbbea15c2b42969486cec74331

woshilapin commented on editoast/core_task/src/envs/simulation/inputs.rs:
```diff
@@ -0,0 +1,266 @@
+                )
+            },
+        );
+        let (simulation_consists, simulation_params, pathfinding_trains): (Vec<_>, Vec<_>, Vec<_>) =
```
Isn’t it possible to keep the iterators when unzipping?

leovalais replied:
Sadly no but it makes sense. Even if we could somehow share the initial iterator to all the unzipped iterators there'd still be an issue. We would be required to iterate all these sub-iterators at the same time. So sadly the source iterator must be consumed completely, thus needing collection.

woshilapin commented on editoast/core_task/src/envs/simulation/request.rs:
```diff
@@ -0,0 +1,129 @@
+        ..
+    }: core_client::pathfinding::PathfindingResultSuccess,
+) -> Result<core_client::simulation::Request, ()> {
+    if constraints.path_items.len() != path_item_positions.len() {
+        tracing::error!(
+            ?sim_consist,
+            ?pf_consist,
+            sim_params = ?params,
+            pf_constraints = ?constraints,
+            waypoint_count = constraints.path_items.len(),
+            computed_path_item_count = path_item_positions.len(),
+            "Core path does not respect pathfinding constraints. Please open a bug report."
+        );
+        return Err(());
+    }
```
I’m not sure that I understand why we do not panic here, and we do panic in other similar piece of code(i.e., invariant not being enforced). We should never end up in this situation anyway, isn’t it?

leovalais replied:
We decided in a preparatory PR about PathfindingEnv that we'd push panics up, closer to the context of why panic is acceptable. I'm fine with panicking here though, especially since it's a private function. Lmk

woshilapin commented on editoast/core_task/src/envs/simulation/inputs.rs:
```diff
@@ -0,0 +1,266 @@
+        pathfinding_inputs: &super::PathfindingEnvInputs<Train>,
+    ) -> impl Iterator<Item = Correlated<Train, Option<super::SimulationKey>>> {
+        self.consists.keys().map(|train| {
+            let pf_key = self.train_input(train, pathfinding_inputs);
```
```suggestion
            let simulation_key = self.train_input(train, pathfinding_inputs);
```

leovalais replied:
22c1be483eef78aded536b507ff32ea640b134d7

woshilapin commented on editoast/core_task/src/envs/simulation/request.rs:
```diff
@@ -0,0 +1,426 @@
+        }
+        margin_boundaries.push(to);
+        margin_values.push(*margin);
+    }
```
We were 3 reviewing this piece of code, and it took us some time to figure out what this `if` was doing. I’m trying to think of a more straightforward to write this (straightforward to understand, I mean). If I’m not wrong, the following is equivalent and show the intent of pushing the first from only once, and checking for the order of values.


```suggestion
    for (range, margin) in params.margins.iter() {
        let from = path_item_positions[range.start];
        let to = path_item_positions[range.end];
        if margin_boundaries.is_empty() {
            margin_boundaries.push(from);
        }
        margin_boundaries.push(to);
        margin_values.push(*margin);
    }
    debug_assert_eq!(margin_boundaries.is_sorted(), "margin boundaries should only distributed along the path, therefore in a sorted order");
```

What do you think?

leovalais replied:
bea004237f95191bad1f66ac6f185b56ed4b5802

Less cryptic and the assertion comment explains the data model well, much better!

woshilapin commented on editoast/core_task/src/envs/simulation/request.rs:
```diff
@@ -0,0 +1,426 @@
+            || margin_boundaries.len() == margin_values.len() + 1
+    );
+
+    let mut schedule = Vec::new();
+    for (i, point) in params.schedule.iter().enumerate() {
+        if let SimulationWaypointSchedule::Stop {
+            arrival_at,
+            stop_for,
+            reception_signal,
+        } = point
+        {
+            let position = path_item_positions[i];
+            schedule.push(core_client::simulation::SimulationScheduleItem {
+                path_offset: position,
+                arrival: *arrival_at,
+                stop_for: *stop_for,
+                reception_signal: *reception_signal,
+            });
+        }
+    }
```
nitpick: The `if` is a **filter**, and therefore, perhaps it would make sense to express it as such. On top of that, it would avoid the `mut` 😉 


```suggestion
    let schedule = params.schedule.iter()
        .enumerate()
        .filter_map(|(i, point)| {
            if let SimulationWaypointSchedule::Stop {
                arrival_at,
                stop_for,
                reception_signal,
            } = point
            {
                let position = path_item_positions[i];
                Some(core_client::simulation::SimulationScheduleItem {
                    path_offset: position,
                    arrival: *arrival_at,
                    stop_for: *stop_for,
                    reception_signal: *reception_signal,
                })
            } else {
                None
            }
        )
        .collect_vec();
```

leovalais replied:
I'm fine with functional style (as you'd expect) but since this function is written in imperative style I thought it'd be best not to mix and match

woshilapin commented on editoast/core_task/src/envs/simulation/request.rs:
```diff
@@ -0,0 +1,426 @@
+
+        let mut builder = SimulationTrain::new(
+            consist(1),
+            pathfinding::test_data::consist(1),
```
```suggestion
            simulation::test_data::consist(1),
            pathfinding::test_data::consist(1),
```

leovalais replied:
ee92c566d257092955b58644a558c63067d302cf

woshilapin commented on editoast/core_task/src/envs/simulation/request.rs:
```diff
@@ -0,0 +1,426 @@
+
+        let mut builder = SimulationTrain::new(
+            consist(1),
+            pathfinding::test_data::consist(1),
```
```suggestion
            simulation::test_data::consist(1),
            pathfinding::test_data::consist(1),
```

leovalais replied:
ee92c566d257092955b58644a558c63067d302cf

woshilapin commented on editoast/core_task/src/envs/simulation/request.rs:
```diff
@@ -0,0 +1,426 @@
+
+    #[test]
+    fn build_request_with_overlapping_ranges() {
+        let path_item_positions = vec![10u64, 20u64, 30u64, 40u64, 50u64];
```
To help the comprehension of this complex test, how about starting at `0u64`, so index corresponds. Because at the moment, index 0 corresponds to 10, index 1 to 20, and so on. I had difficulty at first to understand the overlapped expected values because of this.

leovalais replied:
good idea 2cc48a7c98063540b411761b9e0b0ce19f4afcaf

woshilapin commented on editoast/core_task/src/envs/simulation/request.rs:
```diff
@@ -0,0 +1,426 @@
+            pf_input,
+        );
+
+        build_request(&core_env, None, &sim_key, path_success).expect_err("should fail");
```
nitpick: Just because it’s a test, and we can of expect assertion at some point.

```suggestion
        assert!(build_request(&core_env, None, &sim_key, path_success).is_err(), "should fail");
```

Even better, since we do return an error, let’s match on it. If the error change in the future, the compiler will force us to update this code. With your proposition, or even mine above, the compiler will happily ignore the change.


```suggestion
        assert_eq!(build_request(&core_env, None, &sim_key, path_success), Err(()));
```

leovalais replied:
e7ec79034f667646c3f77c37abe84e8cc49a7674

better indeed

woshilapin commented on editoast/core_task/src/envs/simulation/request.rs:
```diff
@@ -0,0 +1,129 @@
+        }
+        margin_boundaries.push(to);
+        margin_values.push(*margin);
+    }
```
We were 3 reviewing this piece of code, and it took us some time to figure out what this `if` was doing. I’m trying to think of a more straightforward to write this (straightforward to understand, I mean). If I’m not wrong, the following is equivalent and show the intent of pushing the first from only once, and checking for the order of values.

leovalais replied:
duplicate of https://github.com/OpenRailAssociation/osrd/pull/14100#discussion_r2597564219, closing

woshilapin commented on editoast/core_task/src/envs/simulation.rs:
```diff
@@ -0,0 +1,550 @@
+/// The error is forwarded by [SimulationOutput::PathfindingFailure].
+#[derive(Debug)]
+#[cfg_attr(test, derive(PartialEq))]
+#[expect(clippy::large_enum_variant)] // yeah it's fine
```
How about a real comment like we already have in other places 😊

leovalais replied:
anything wrong with my comment??

8c070f94412ea740c433d18d6318095f4aa10fd4

woshilapin commented on editoast/core_task/src/envs/simulation.rs:
```diff
@@ -0,0 +1,310 @@
+where
+    Train: Clone + Hash + Eq + Send + Sync + 'static,
+{
+    pub fn new(core_env: CoreEnv, electrical_profiles_set_id: Option<u64>) -> Self {
```
nitpick: As often, I’m not fond of optional parameter as function arguments. The caller don’t have a clue about what is going on if `None`. One way to do differently is the following:

```rust
    pub fn new(core_env: CoreEnv) -> Self {

        Self {
            pathfinding_env: PathfindingEnv::new(core_env),
            inputs: SimulationInputs {
                electrical_profiles_set_id: None,
                consists: Default::default(),
                parameters: Default::default(),
            },
        }
    }

   pub fn new_with_electrical_profile_set(core_env: CoreEnv, electrical_profiles_set_id: u64) -> Self {
Comment
        Self {
            pathfinding_env: PathfindingEnv::new(core_env),
            inputs: SimulationInputs {
                electrical_profiles_set_id: Some(electrical_profiles_set_id),
                consists: Default::default(),
                parameters: Default::default(),
            },
        }
    }

leovalais replied:
eb794f86bbeda2462a9bae474447f19593bd858e

woshilapin commented on editoast/core_task/src/envs/simulation/inputs.rs:
```diff
@@ -0,0 +1,266 @@
+    pub(super) fn iter(
+        &self,
+        pathfinding_inputs: &super::PathfindingEnvInputs<Train>,
+    ) -> impl Iterator<Item = Correlated<Train, Option<super::SimulationKey>>> {
```
Do we have a use for returning a `Correlated` if `SimulationKey` is not present? Instead of yielding `Correlated<Train, Option<super::SimulationKey>>`, shouldn’t we yield `Option<Correlated<Train, super::SimulationKey>>` instead? Or maybe panic if it happens (instead of panicking [here](https://github.com/OpenRailAssociation/osrd/pull/14100/files#diff-1f9f30ae299d3979c4c1c191ed35f88222a7833e44698ca9d9d181bdbb889b2eR145)).

leovalais replied:
Likewise it's about pushing panics up. But I'm fine with panicking here as well. Lmk

woshilapin commented on editoast/core_task/src/envs/simulation.rs:
```diff
@@ -0,0 +1,550 @@
+    /// Trains sharing the same pathfinding key (= inputs), therefore having the same path,
+    /// do not necessarily share the same simulation key (simulation inputs differ).
+    fn simulation_keys_of_pathfinding_key(
+        &self,
+        pathfinding_key: &pathfinding::Input,
+    ) -> Option<impl Iterator<Item = SimulationKey>> {
+        Some(
+            self.pathfinding_runner
+                .input_train_set(pathfinding_key)?
+                .into_iter()
+                .filter_map(|train| {
+                    self.simulation_inputs
+                        .train_input(&train, &self.pathfinding_runner.pathfinding_inputs)
+                }),
+        )
+    }
```
Isn’t this equivalent and maybe a bit more idiomatic (`Option<Iterator>` feels weird). Or is it really useful to make a distinction between the 2 cases?

```rust
    fn simulation_keys_of_pathfinding_key(
        &self,
        pathfinding_key: &pathfinding::Input,
    ) -> impl Iterator<Item = SimulationKey> {
        self.pathfinding_runner
            .input_train_set(pathfinding_key)?
            .into_iter()
            .flatten()
            .filter_map(|train| {
                self.simulation_inputs
                    .train_input(&train, &self.pathfinding_runner.pathfinding_inputs)
            }),
    )

```

It might actually be fine to also panic here instead of returning an `Option` (since this is actually what is done anyway every time this function is called).

leovalais replied:
that's indeed better, no need to panic anymore even 01d3a061317077d3bff56a41c0378c174f9e9b3d

woshilapin commented on editoast/core_task/src/envs/simulation.rs:
```diff
@@ -0,0 +1,310 @@
+                            return_tx
+                                .unbounded_send(Correlated::new(
+                                    sim_key,
+                                    Ok(SimulationOutput::PathfindingFailure(failure.clone())),
```
Isn’t `failure` owned by the current match binding? Why do we need to clone here? (pure curiosity, feel free to resolve that one)

leovalais replied:
This is in a loop, so that'd move the `failure` in the first iteration.

woshilapin commented on editoast/core_task/src/envs/simulation.rs:
```diff
@@ -0,0 +1,550 @@
+
+        let vk = cache::Client::new_mock(
+            vec![
+                mock_mget(vec![
+                    (simenv.pathfinding_env.key(1), Some(path(1, 4))),
+                    (simenv.pathfinding_env.key(2), Some(path(2, 4))),
+                ]),
+                mock_mget(vec![
+                    (simenv.cache_key(1), Some(simulation_success(1))),
+                    (simenv.cache_key(2), None),
+                ]),
+                cache::MockCmd::new(
```
I believe `PathfindingEnv::key()` is the equivalent of `SimulationEnv::cache_key()`. Probably a good idea to name them similarly? (does a trait make sense to ensure that?) I’m not sure which one I prefer, objectively, it’s a key, that happens to be used for caching, but, when I read `key()`, I wondered a moment what it was, but then I read `cache_key()` and it all made sense immediately… so I’d say the latter is probably the best?

leovalais replied:
Yes it's one of the naming disparities I was asking about. cache_key is also my preference.

woshilapin commented on editoast/core_task/src/envs/simulation/request.rs:
```diff
@@ -0,0 +1,426 @@
+    }
+
+    #[test]
+    fn build_request_error_schedule_path_mismatch() {
```
I have seen the comment in the commit description explaining what each test does, it’s great to quickly grasp what the test is about… but I believe it’s not in the right place for 2 reasons:
- when I try to understand a test, I first look at the test, not the commit description
- I was under the impression that the commit are not there to stay (only help split the PR for review) so the commit description will probably go away on squash

Maybe put the description as comment above each test?

leovalais replied:
I thought the test names were good enough, I've added the comments in b210f59cf92fae227ab01d59ce8b4d51db80b675

woshilapin commented on editoast/core_client/src/simulation.rs:
```diff
@@ -237,19 +255,9 @@ pub struct SpeedLimitProperties {
 }
 
-#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, ToSchema)]
-pub struct SimulationPowerRestrictionRange {
```
Why is this going away? (if unused, could have been another PR)

leovalais replied:
Oops. It was supposed to be dedicated commit. I failed my splitting...

woshilapin reviewed (commented):
I know we had multiple peer-review on that one, some of my comments were done before, some after. Let’s address the obvious one and discuss the other one if needed.

woshilapin commented on editoast/core_task/src/envs/simulation/inputs.rs:
```diff
@@ -0,0 +1,266 @@
+where
+    Train: Clone + std::hash::Hash + Eq + Send + Sync + 'static,
+{
+    pub(super) electrical_profiles_set_id: Option<u64>,
```
Not ultra sure but, it is **one** electrical profile, isn’t it? At least, that’s how it’s named in `SimulationBatchForm` already so we should make sure it’s homogeneous everywhere.

```suggestion
    pub(super) electrical_profile_set_id: Option<u64>,
```

leovalais replied:
c7f991bec07aa44a61ab56b1b84553b85c813780

leovalais committed:
editoast: core_client: remove unused definition

Signed-off-by: Leo Valais <leo.valais97@gmail.com>

woshilapin committed:
editoast: use schedule item labels for 'push_waypoint'

The API let the caller define labels for each waypoint, then reuse them
for setting up the power restrictions and margins.


Signed-off-by: Jean SIMARD <woshilapin@tuziwo.info>

MartinBourbier reviewed (approved):
LGTM 🦅
